src/output-json-anomaly.c

```diff
@@ -134,6 +139,129 @@ static int AnomalyJson(ThreadVars *tv, JsonAnomalyLogThread *aft, const Packet *
     return TM_ECODE_OK;
 }
 
+static int AnomalyAppLayerDecoderEventJson(JsonAnomalyLogThread *aft,
+                        const Packet *p, AppLayerDecoderEvents *decoder_events,
+                        bool is_applayer, const char *layer, uint64_t tx_id)
+{
+    const char *alprotoname;
+    alprotoname = AppLayerGetProtoName(p->flow->alproto);
+
+    for (int i = 0; i < decoder_events->cnt; i++) {
+        MemBufferReset(aft->json_buffer);
+
+        json_t *js = CreateJSONHeader(p, LOG_DIR_PACKET, ANOMALY_EVENT_TYPE);
+        if (unlikely(js == NULL)) {
+            return TM_ECODE_OK;
+        }
+
+        json_t *ajs = json_object();
+        if (unlikely(ajs == NULL)) {
+            json_decref(js);
+            return TM_ECODE_OK;
+        }
+
+        JsonFiveTuple((const Packet *)p, LOG_DIR_PACKET, js);
+        JsonAddCommonOptions(&aft->json_output_ctx->cfg, p, p->flow, js);
+
+        /* Use app layer proto name if available */
+        if (alprotoname) {
+            json_object_set_new(ajs, "alproto", json_string(alprotoname));
+        } else {
+            json_object_set_new(ajs, "alproto",
+                p->flow ? json_integer(p->flow->alproto) : json_string("unknown"));
+        }
+
+        const char *event_name = NULL;
+        uint8_t event_code = decoder_events->events[i];
+        AppLayerEventType event_type;
+        int r;
+        if (is_applayer) {
+            r = AppLayerGetEventInfoById(event_code, &event_name, &event_type);
+        } else {
+            r = AppLayerParserGetEventInfoById(p->flow->proto, p->flow->alproto,
+                                               event_code, &event_name, &event_type);
+        }
+        if (r == 0) {
+            json_object_set_new(ajs, "type",
+                        json_string(event_type == APP_LAYER_EVENT_TYPE_TRANSACTION ?
+                                    "transaction" : "packet"));
+            json_object_set_new(ajs, "event", json_string(event_name));
+        } else {
+            json_object_set_new(ajs, "type", json_string("unknown"));
+            json_object_set_new(ajs, "code", json_integer(event_code));
+        }
+
+        /* log event number (and total */
+        uint32_t offset = 0;
+        char event_no_buf[32];
+        PrintBufferData(event_no_buf, &offset, 32, "%d (of %d)",
```

Comment:
is there a reason for not using snprintf here?

Reply:
No; will use that here and in first anomaly logger.